cpp/CMakeLists.txt

```diff
@@ -251,22 +252,9 @@ target_compile_definitions(raft::raft INTERFACE $<$<BOOL:${RAFT_NVTX}>:NVTX_ENAB
 endif()
 
 # ##################################################################################################
-# * raft_compiled ------------------------------------------------------------ TODO: Currently, this
-#   package also contains the 'random' namespace (for rmat logic) We couldn't get this to work
-#   properly due to strange CI failures as noticed in the PR#778. In the long term, we should rename
-#   this package to `raft_compiled` in order to have a single pre-compiled raft package for those
-#   who need it.
-add_library(raft_compiled INTERFACE)
-
-if(TARGET raft_compiled AND (NOT TARGET raft::compiled))
-  add_library(raft::compiled ALIAS raft_compiled)
-endif()
-
-set_target_properties(raft_compiled PROPERTIES EXPORT_NAME compiled)
+# * raft_compiled ------------------------------------------------------------
 
-if(RAFT_COMPILE_LIBRARY)
-  add_library(
-    raft_lib
+set(RAFT_COMPILED_SOURCE_FILES
```

Comment:
With the existing implementation we will compile each source file twice when static has been enabled.

What we should do is have a raft_internal or raft_obj library which is of type `add_library( OBJECT)`. This would ensure that we only compile each source file once, even when static is enabled.

Reply:
Rough diff of the proposed changes:

```
diff --git a/cpp/CMakeLists.txt b/cpp/CMakeLists.txt
index 3d9f8a3b..b068b872 100644
--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -430,40 +430,54 @@ endif()
 
 set_target_properties(raft_compiled PROPERTIES EXPORT_NAME compiled)
 
-if(RAFT_COMPILE_LIBRARY)
-  add_library(raft_lib ${RAFT_COMPILED_SOURCE_FILES})
+if(RAFT_COMPILE_LIBRARY OR RAFT_COMPILE_STATIC_LIBRARY)
+  add_library(raft_objs OBJECT ${RAFT_COMPILED_SOURCE_FILES})
   set_target_properties(
-    raft_lib
-    PROPERTIES OUTPUT_NAME raft
-               BUILD_RPATH "\$ORIGIN"
-               INSTALL_RPATH "\$ORIGIN"
-               CXX_STANDARD 17
+    raft_objs
+    PROPERTIES CXX_STANDARD 17
                CXX_STANDARD_REQUIRED ON
                CUDA_STANDARD 17
                CUDA_STANDARD_REQUIRED ON
-               POSITION_INDEPENDENT_CODE ON
-               INTERFACE_POSITION_INDEPENDENT_CODE ON
-  )
+               POSITION_INDEPENDENT_CODE ON)
 
   target_link_libraries(
-    raft_lib
+    raft_objs
     PUBLIC raft::raft
            ${RAFT_CTK_MATH_DEPENDENCIES} # TODO: Once `raft::resources` is used everywhere, this
                                          # will just be cublas
            $<TARGET_NAME_IF_EXISTS:OpenMP::OpenMP_CXX>
   )
+
   target_compile_options(
-    raft_lib PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:${RAFT_CXX_FLAGS}>"
+    raft_objs PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:${RAFT_CXX_FLAGS}>"
                      "$<$<COMPILE_LANGUAGE:CUDA>:${RAFT_CUDA_FLAGS}>"
   )
 
   # RAFT_COMPILED is set during compilation of libraft.so as well as downstream libraries (due to
   # "PUBLIC")
-  target_compile_definitions(raft_lib PUBLIC "RAFT_COMPILED")
+  target_compile_definitions(raft_objs PUBLIC "RAFT_COMPILED")
 
   # RAFT_EXPLICIT_INSTANTIATE_ONLY is set during compilation of libraft.so (due to "PRIVATE")
-  target_compile_definitions(raft_lib PRIVATE "RAFT_EXPLICIT_INSTANTIATE_ONLY")
+  target_compile_definitions(raft_objs PRIVATE "RAFT_EXPLICIT_INSTANTIATE_ONLY")
+endif()
+
+if(RAFT_COMPILE_LIBRARY)
+  add_library(raft_lib SHARED $<TARGET_OBJECTS:raft_objs>)
+  set_target_properties(
+    raft_lib
+    PROPERTIES OUTPUT_NAME raft
+               BUILD_RPATH "\$ORIGIN"
+               INSTALL_RPATH "\$ORIGIN"
+               INTERFACE_POSITION_INDEPENDENT_CODE ON
+  )
 
+  target_link_libraries(
+    raft_lib
+    PUBLIC raft::raft
+           ${RAFT_CTK_MATH_DEPENDENCIES} # TODO: Once `raft::resources` is used everywhere, this
+                                         # will just be cublas
+           $<TARGET_NAME_IF_EXISTS:OpenMP::OpenMP_CXX>
+  )
   # ensure CUDA symbols aren't relocated to the middle of the debug build binaries
   target_link_options(raft_lib PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/fatbin.ld")
 
@@ -478,7 +492,7 @@ target_link_libraries(raft_compiled INTERFACE raft::raft $<TARGET_NAME_IF_EXISTS
 # ##################################################################################################
 # * raft_compiled_static----------------------------------------------------------------------------
 
-add_libraft(raft_compiled_static INTERFACE)
+add_library(raft_compiled_static INTERFACE)
 
 if(TARGET raft_compiled_static AND (NOT TARGET raft::compiled_static))
   add_library(raft::compiled_static ALIAS raft_compiled_static)
@@ -486,17 +500,12 @@ endif()
 set_target_properties(raft_compiled_static PROPERTIES EXPORT_NAME compiled_static)
 
 if(RAFT_COMPILE_STATIC_LIBRARY)
-  add_library(raft_lib_static ${RAFT_COMPILED_SOURCE_FILES})
+  add_library(raft_lib_static STATIC $<TARGET_OBJECTS:raft_objs>)
   set_target_properties(
     raft_lib_static
     PROPERTIES OUTPUT_NAME raft
                BUILD_RPATH "\$ORIGIN"
                INSTALL_RPATH "\$ORIGIN"
-               CXX_STANDARD 17
-               CXX_STANDARD_REQUIRED ON
-               CUDA_STANDARD 17
-               CUDA_STANDARD_REQUIRED ON
-               POSITION_INDEPENDENT_CODE ON
                INTERFACE_POSITION_INDEPENDENT_CODE ON
   )
 
@@ -507,18 +516,6 @@ if(RAFT_COMPILE_STATIC_LIBRARY)
            # will just be cublas
            $<TARGET_NAME_IF_EXISTS:OpenMP::OpenMP_CXX>
   )
-  target_compile_options(
-    raft_lib_static PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:${RAFT_CXX_FLAGS}>"
-                            "$<$<COMPILE_LANGUAGE:CUDA>:${RAFT_CUDA_FLAGS}>"
-  )
-
-  # RAFT_COMPILED is set during compilation of libraft.so as well as downstream libraries (due to
-  # "PUBLIC")
-  target_compile_definitions(raft_lib_static PUBLIC "RAFT_COMPILED")
-
-  # RAFT_EXPLICIT_INSTANTIATE_ONLY is set during compilation of libraft.so (due to "PRIVATE")
-  target_compile_definitions(raft_lib_static PRIVATE "RAFT_EXPLICIT_INSTANTIATE_ONLY")
-
   # ensure CUDA symbols aren't relocated to the middle of the debug build binaries
   target_link_options(raft_lib_static PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/fatbin.ld")
 
```